### PR TITLE
Reduction OP

### LINKED ISSUE #1075: Update Reduction operator

### PR DESCRIPTION
**Pull request by lambda7xx**
**Description of changes:**
- upate the Reduction OP

**Related Issues:**

Linked Issues:
- Issue #1075

Issues closed by this PR:
- Closes #1075

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)

### CODE REVIEW

**Review by wmdi — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @lambda7xx, @lockshaw, and @reyna-abhyankar)

___
*[`lib/runtime/src/ops/reduction.cc` line 19 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1120#-NfHjZTu5MYPV2exG8rM:-NfHjZTu5MYPV2exG8rN:b-2n1f7f) ([raw file](https://github.com/flexflow/FlexFlow/blob/d7e471693e3726f2625e26de7a0de0434a682e71/lib/runtime/src/ops/reduction.cc#L19)):*
> ```cpp
> #include "kernels/reduction_kernels.h"
> #include "op-attrs/get_output_shape.h"
> #include "utils/exception.decl.h"
> ```

`decl.h` files should not be included here.

**Review by wmdi — CHANGES_REQUESTED** [review]
> Reviewed 2 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)* status: 2 of 3 files reviewed, 1 unresolved discussion (waiting on @lambda7xx, @lockshaw, and @reyna-abhyankar)

___
*[`lib/runtime/src/ops/reduction.cc` line 19 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1120#-NfHjZTu5MYPV2exG8rM:-NfN299f7lMiHupPfH3R:b-iiwhkp) ([raw file](https://github.com/flexflow/FlexFlow/blob/d7e471693e3726f2625e26de7a0de0434a682e71/lib/runtime/src/ops/reduction.cc#L19)):*
<details><summary><i>Previously, wmdi (Mengdi Wu) wrote…</i></summary><blockquote>

`decl.h` files should not be included here.
</blockquote></details>

I also noticed a few other files outside utils including`.decl.h`. Remember to update them as well.

**Comment by lambda7xx** [comment]
*[`lib/runtime/src/ops/reduction.cc` line 19 at r1](https://reviewable.io/reviews/flexflow/FlexFlow/1120#-NfHjZTu5MYPV2exG8rM:-NfNFDgj3ETEizqUnWLW:b-291hgk) ([raw file](https://github.com/flexflow/FlexFlow/blob/d7e471693e3726f2625e26de7a0de0434a682e71/lib/runtime/src/ops/reduction.cc#L19)):*
<details><summary><i>Previously, lambda7xx (Lambda(Xiaoxiang) Shi ) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

currently, other files like `contaners.h` use .`decl.h` is not my side. I will check my other pr

File: lib/runtime/src/ops/reduction.cc [hunk 1]
```diff
@@ -47,206 +49,117 @@ ReductionParams Reduction::get_params() const {
...
-  runtime->execute_index_space(ctx, launcher);
+static optional<float> forward_task_impl(TaskArgumentAccessor const &acc) {
+  ProfilingSettings profiling_settings =
+      acc.get_argument<ProfilingSettings>(PROFILING);
+
+  auto input = acc.get_tensor<Permissions::RO>(INPUT);
+  auto output = acc.get_tensor<Permissions::WO>(OUTPUT);
+  auto attrs = acc.get_argument<ReductionAttrs>(ATTRS);
+  // Note: how to get num_replicas?
+  size_t num_replicas = attrs.reduction_dim;
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/reduction.cc:79]:
```suggestion
  size_t num_replicas = attrs.reduction_degree;
```

I think degree is correct here, as it is the number of partitions along the reduction dimension? But maybe @lockshaw can confirm

Comment (reyna-abhyankar) [reply lib/runtime/src/ops/reduction.cc:79]: However, if this is wrong, you can add a slot for `INPUT_PARALLEL_TENSOR_SHAPE` and then do `acc.get_argument<ParallelTensorShape>(INPUT_PARALLEL_TENSOR_SHAPE)` and get num replicas from there.

File: lib/runtime/src/ops/reduction.cc [hunk 2]
```diff
@@ -47,206 +49,117 @@ ReductionParams Reduction::get_params() const {
...
-  float *input_grad_ptr = helperGetTensorPointerWO<float>(
-      regions[1], task->regions[1], FID_DATA, ctx, runtime);
-
-  backward_kernel<float>(
-      output_grad_ptr, input_grad_ptr, output_grad_domain.get_volume());
+template <>
+void register_task<REDUCTION_FWD_TASK_ID>() {
+  OpTaskSignature fwd(OpTaskType::FWD);
+
+  fwd.add_arg_slot<ProfilingSettings>(PROFILING);
```
Comment (reyna-abhyankar) [inline lib/runtime/src/ops/reduction.cc:149]:
```suggestion
  fwd.add_arg_slot<ProfilingSettings>(PROFILING);
  fwd.add_arg_slot<ReductionAttrs>(ATTRS);
```

**Review by reyna-abhyankar — CHANGES_REQUESTED** [review]

File: lib/runtime/src/ops/reduction.cc (hunk 1 above)
Comment (reyna-abhyankar) [reply lib/runtime/src/ops/reduction.cc:72]: bump @lockshaw

File: lib/runtime/src/ops/reduction.cc (hunk 2 above)
Comment (reyna-abhyankar) [reply lib/runtime/src/ops/reduction.cc:142]: bump @lambda7xx

**Review by reyna-abhyankar — CHANGES_REQUESTED** [review]

**Review by lambda7xx — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)* status: 2 of 3 files reviewed, 3 unresolved discussions (waiting on @lockshaw, @reyna-abhyankar, and @wmdi)

___
*[`lib/runtime/src/ops/reduction.cc` line 79 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1120#gh-1341911349:-NgzKxnlC5-ZpUpz8Jpm:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/d4ae7ce876a782866e42c33f12dab925069af2fe/lib/runtime/src/ops/reduction.cc#L79)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

bump @lockshaw 
</blockquote></details>

Done.
___
*[`lib/runtime/src/ops/reduction.cc` line 149 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1120#gh-1341911915:-NgzKwUmDX9Eyvw8nAS4:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/d4ae7ce876a782866e42c33f12dab925069af2fe/lib/runtime/src/ops/reduction.cc#L149)):*
<details><summary><i>Previously, reyna-abhyankar (Reyna Abhyankar) wrote…</i></summary><blockquote>

bump @lambda7xx 
</blockquote></details>

Done.

**Review by wmdi — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)* status: 2 of 3 files reviewed, 2 unresolved discussions (waiting on @lambda7xx, @lockshaw, and @reyna-abhyankar)

**Review by wmdi — DISMISSED** [review]
> Reviewed 1 of 1 files at r4.
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)* status: all files reviewed, 2 unresolved discussions (waiting on @lockshaw and @reyna-abhyankar)

File: lib/runtime/src/ops/reduction.cc (hunk 1 above)
Comment (lockshaw) [reply lib/runtime/src/ops/reduction.cc:72]: Yeah I agree @reyna-abhyankar

**Review by lambda7xx — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1120)* status: all files reviewed, 2 unresolved discussions (waiting on @reyna-abhyankar)

___
*[`lib/runtime/src/ops/reduction.cc` line 79 at r2](https://reviewable.io/reviews/flexflow/FlexFlow/1120#gh-1341911349:-Nhd0HRj5wJ-yn8_lUrh:b-896fix) ([raw file](https://github.com/flexflow/FlexFlow/blob/d4ae7ce876a782866e42c33f12dab925069af2fe/lib/runtime/src/ops/reduction.cc#L79)):*
<details><summary><i>Previously, lockshaw (Colin Unger) wrote…</i></summary><blockquote>

Yeah I agree @reyna-abhyankar 
</blockquote></details>

Done.

**Review by reyna-abhyankar — APPROVED** [review]

**Pull request merged**